interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,310 @@
+		}
+	}
+	return list, nil
+}
+
+func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error) error {
+	mountAttr, ok := plug.Lookup("mount")
+	if !ok {
+		return nil
+	}
+	mounts, ok := mountAttr.([]interface{})
+	if !ok {
+		return fmt.Errorf(`mount-control "mount" attribute must be an array`)
+	}
+
+	for _, m := range mounts {
+		mount, ok := m.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf(`mount-control "mount" attribute is not a dictionary`)
+		}
+
+		what, ok := mount["what"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount::what" must be a string`)
+		}
+
+		where, ok := mount["where"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount::where" must be a string`)
+		}
+
+		persistent := false
+		persistentValue, ok := mount["persistent"]
+		if ok {
+			if persistent, ok = persistentValue.(bool); !ok {
+				return fmt.Errorf(`mount-control "mount::persistent" must be a boolean`)
+			}
+		}
+
+		types, err := parseStringList(mount, "type")
+		if err != nil {
+			return err
+		}
+
+		options, err := parseStringList(mount, "options")
+		if err != nil {
+			return err
+		}
+
+		// TODO: parse other fields
+
+		mountInfo := &MountInfo{
+			what:       what,
+			where:      where,
+			persistent: persistent,
+			types:      types,
+			options:    options,
+		}
+
+		if err := fn(mountInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateWhatAttr(what string) error {
+	if what[0] != '/' {
+		return fmt.Errorf(`mount-control "what" attribute must be an absolute path`)
+	}
+
+	if _, err := utils.NewPathPattern(what); err != nil {
+		return fmt.Errorf(`mount-control "what" attribute error: %v`, err)
+	}
+
+	// TODO add more checks?
+	return nil
+}
+
+func validateWhereAttr(where string) error {
+	if !whereRegexp.MatchString(where) {
+		return fmt.Errorf(`mount-control "where" attribute is invalid`)
+	}
+
+	if _, err := utils.NewPathPattern(where); err != nil {
+		return fmt.Errorf(`mount-control "where" attribute error: %v`, err)
+	}
+
+	// TODO add more checks?
+	return nil
+}
+
+func validateMountOptions(options []string) error {
+	for _, o := range options {
+		if !strutil.ListContains(allowedMountOptions, o) {
+			return fmt.Errorf(`mount-control option unrecognized or forbidden: %q`, o)
+		}
+	}
+	return nil
+}
+
+func validateMountInfo(mountInfo *MountInfo) error {
+	if err := validateWhatAttr(mountInfo.what); err != nil {
+		return err
+	}
+
+	if err := validateWhereAttr(mountInfo.where); err != nil {
+		return err
+	}
+
+	if err := validateMountOptions(mountInfo.options); err != nil {
+		return err
+	}
+
+	// TODO: validate mount type
+
+	return nil
+}
+
+func (iface *mountControlInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	// The systemd.ListMountUnits() method works by issuing the command
+	// "systemctl show *.mount", but globbing was only added in systemd v209.
+	if err := systemd.EnsureAtLeast(209); err != nil {
```

Comment:
I think we might want to check for this only on connection and not simply plug existence, so that would be BeforeConnectPlug. Maybe @anonymouse64 has also an opinion on this